### PR TITLE
[AIRFLOW-970] Latest runs on homepage should load async and in batch

### PR DESCRIPTION
**Pull request by saguziel**
Please accept this PR that addresses the following issues:
- https://issues.apache.org/jira/browse/AIRFLOW-970

Testing Done:
- Opened UI, existing tests, and identical HTML

This causes our render time to go from 6s to 2s.

@aoen

### CODE REVIEW

File: airflow/www/views.py [hunk 1]
```diff
@@ -1084,6 +1084,28 @@ def blocked(self):
...
+        DR = models.DagRun
+        subquery = (
+            session.query(DR.dag_id, sqla.func.max(DR.execution_date).label('execution_date'))
+            .filter(DR.state == State.RUNNING)
+            .group_by(DR.dag_id)
+            .subquery()
+        )
+        dagruns = session.query(DR).join(subquery, and_(DR.dag_id == subquery.c.dag_id, DR.execution_date == subquery.c.execution_date)).all()
+        payload = []
+        for dagrun in dagruns:
```
Comment (aoen) [inline airflow/www/views.py]: Let's be defensive and shield against dagruns that don't have these properties defined.

File: airflow/www/views.py [hunk 2]
```diff
@@ -1084,6 +1084,28 @@ def blocked(self):
             })
         return wwwutils.json_response(payload)
 
+    @expose('/latest_dag_runs')
+    @login_required
+    def latest_dag_runs(self):
```
Comment (bolkedebruin) [inline airflow/www/views.py]: Please move this out from views. Ie. put the business logic into api and remove any database access from views.py. This makes it easier to unit test and to expose it by Rest API later.

**Review by bolkedebruin — CHANGES_REQUESTED** [review]
Please add unit tests and be a more elaborate in your commit message :).

File: airflow/www/views.py (hunk 2 above)
Comment (saguziel) [reply airflow/www/views.py]: @bolkedebruin is the business logic okay as a class method in models.Dag or models.DagRun or are we moving in the direction of putting things like this into API

**Comment by saguziel** [comment]
@bolkedebruin PTAL

File: airflow/www/views.py (hunk 2 above)
Comment (bolkedebruin) [reply airflow/www/views.py]: I have no particular preference tbh. API feels a bit more natural as it is not just Data Access, but with some extra logic.

File: airflow/www/views.py [hunk 3]
```diff
@@ -1084,6 +1084,26 @@ def blocked(self):
...
+        """Returns the latest running DagRun for each DAG formatted for the UI. """
+        session = settings.Session()
```
Comment (bolkedebruin) [inline airflow/www/views.py]: I don't think "views" should have any db access on their own and we should move away from that. Please don't open a session here and leave that to the API logic

**Comment by saguziel** [comment]
@bolkedebruin unless we want private APIs there I think we should leave it. This has special logic that's not really useful unless for the homepage

**Comment by bolkedebruin** [comment]
@saguziel we will have private API's (Task to API communication) eventually, but we can always move stuff. As mentioned I am not particularly opinionated here. 

Can you take care of removing the the DB access from the function in views?

**Comment by saguziel** [comment]
The db access is contained entirely within models now

**Comment by bolkedebruin** [comment]
Session is still opened in views? Also please verify your commit message against the guidelines pls (subject imperative and subject length)

**Comment by saguziel** [comment]
bump @bolkedebruin

File: airflow/www/api/experimental/endpoints.py [hunk 4]
```diff
@@ -110,3 +112,24 @@ def task_info(dag_id, task_id):
...
     fields = {k: str(v) for k, v in vars(task).items() if not k.startswith('_')}
     return jsonify(fields)
+
+
+@api_experimental.route('/latest_runs', methods=['GET'])
+@requires_authentication
+def latest_dag_runs():
+    """Returns the latest running DagRun for each DAG formatted for the UI. """
+    from airflow.models import DagRun
+    session = settings.Session()
```
Comment (bolkedebruin) [inline airflow/www/api/experimental/endpoints.py]: Please don't open DB sessions in views.py

Comment (saguziel) [reply airflow/www/api/experimental/endpoints.py]: @bolkedebruin this isn't views.py

Comment (bolkedebruin) [reply airflow/www/api/experimental/endpoints.py]: Ah sorry. In this case however the DB access should be move to api/common/experimental. See the other examples. Or just use the provided session and don't open a session at all here.

**Comment by saguziel** [comment]
bump again

**Pull request closed without merging**